plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/decoders/ScaledFloatDecoder.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.elasticsearch.decoders;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.trino.plugin.elasticsearch.DecoderDescriptor;
+import io.trino.spi.TrinoException;
+import io.trino.spi.block.BlockBuilder;
+import org.elasticsearch.search.SearchHit;
+
+import java.util.function.Supplier;
+
+import static io.trino.spi.StandardErrorCode.TYPE_MISMATCH;
+import static io.trino.spi.type.DoubleType.DOUBLE;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class ScaledFloatDecoder
+        implements Decoder
+{
+    private final String path;
+    private final double scale;
+
+    public ScaledFloatDecoder(String path, double scale)
+    {
+        this.path = requireNonNull(path, "path is null");
+        this.scale = scale;
+    }
+
+    @Override
+    public void decode(SearchHit hit, Supplier<Object> getter, BlockBuilder output)
+    {
+        Object value = getter.get();
+        if (value == null) {
+            output.appendNull();
+            return;
+        }
+
+        long decoded;
+        if (value instanceof Number) {
+            decoded = ((Number) value).longValue();
+        }
+        else if (value instanceof String) {
+            try {
+                decoded = Long.parseLong((String) value);
+            }
+            catch (NumberFormatException e) {
+                throw new TrinoException(TYPE_MISMATCH, format("Cannot parse value for field '%s' as DOUBLE: %s", path, value));
+            }
+        }
+        else {
+            throw new TrinoException(TYPE_MISMATCH, format("Expected a numeric value for field %s of type DOUBLE: %s [%s]", path, value, value.getClass().getSimpleName()));
+        }
+
+        DOUBLE.writeDouble(output, decoded / scale);
```

Comment:
Based on the [docs](https://www.elastic.co/guide/en/elasticsearch/reference/master/number.html): 

> For floating-point types, it is often more efficient to store floating-point data into an integer using a scaling factor, which is what the scaled_float type does under the hood. For instance, a price field could be stored in a scaled_float with a scaling_factor of 100. All APIs would work as if the field was stored as a double, but under the hood Elasticsearch would be working with the number of cents, price*100, which is an integer. This is mostly helpful to save disk space since integers are way easier to compress than floating points. scaled_float is also fine to use in order to trade accuracy for disk space. For instance imagine that you are tracking cpu utilization as a number between 0 and 1. It usually does not matter much whether cpu utilization is 12.7% or 13%, so you could use a scaled_float with a scaling_factor of 100 in order to round cpu utilization to the closest percent in order to save space.

and

> The scaling factor to use when encoding values. Values will be multiplied by this factor at index time and rounded to the closest long value. For instance, a scaled_float with a scaling_factor of 10 would internally store 2.34 as 23 and all search-time operations (queries, aggregations, sorting) will behave as if the document had a value of 2.3. High values of scaling_factor improve accuracy but also increase space requirements. This parameter is required.

wouldn't writing `12345678` with 100 SF be stored as `1234567800` and returned as `12345678` API? Why do we need to divide it again by SF?

Reply:
Yes, you're right. I was under the mistaken impression that the scale factor mattered to how the client interpreted the value, but it's only an internal setting.